From 8d446b0682fa93eb700c4d2396b935667fdf4bcc Mon Sep 17 00:00:00 2001
From: Sandor Yu <Sandor.yu@nxp.com>
Date: Fri, 15 Nov 2019 10:45:23 +0800
Subject: [PATCH] MLK-22868: DRM: imx: hdmi: Fix kernel panic issue when system
 suspend

commit 314289dadfb347b8c804bc20ee568622aeb4320a from
https://source.codeaurora.org/external/imx/linux-imx.git

i.MX8MQ will kernel panic when system enter suspend;
[   42.371412] Call trace:
[   42.371417]  __flush_work+0x214/0x238
[   42.371422]  __cancel_work_timer+0x128/0x198
[   42.371427]  cancel_delayed_work_sync+0x10/0x18
[   42.371432]  imx_hdcp_disable+0x3c/0x68
[   42.371436]  imx_hdp_imx_encoder_disable+0x1c/0x60
[   42.371447]  drm_atomic_helper_commit_modeset_disables+0xdc/0x3c8

It is cause by function imx_hdcp_disable try to cancel unscheduled thread.
Add hdcp enable value check before call cancel_delayed_work_sync
to avoid kernel panic.

Signed-off-by: Sandor Yu <Sandor.yu@nxp.com>
(cherry picked from commit 706d7860d8f6bc957e8a4a180ce194414e696ecf)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-hdcp-private.h | 1 +
 drivers/gpu/drm/imx/hdp/imx-hdcp.c         | 7 ++++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/hdp/imx-hdcp-private.h b/drivers/gpu/drm/imx/hdp/imx-hdcp-private.h
index a8aa12d8f865..41a68185ba38 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdcp-private.h
+++ b/drivers/gpu/drm/imx/hdp/imx-hdcp-private.h
@@ -21,5 +21,6 @@ struct imx_hdcp {
 	struct work_struct prop_work;
 	u8 bus_type;
 	u8 config;
+	u8 enable;
 };
 #endif
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdcp.c b/drivers/gpu/drm/imx/hdp/imx-hdcp.c
index 4f10737953d9..247e4a5c1799 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdcp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdcp.c
@@ -581,6 +581,7 @@ int imx_hdcp_enable(struct imx_hdp *hdp)
 	schedule_work(&hdp->hdcp.prop_work);
 	schedule_delayed_work(&hdp->hdcp.check_work,
 			      DRM_HDCP_CHECK_PERIOD_MS);
+	hdp->hdcp.enable = 1;
 out:
 	mutex_unlock(&hdp->hdcp.mutex);
 	return ret;
@@ -596,7 +597,11 @@ int imx_hdcp_disable(struct imx_hdp *hdp)
 		ret = _imx_hdcp_disable(hdp);
 	}
 	mutex_unlock(&hdp->hdcp.mutex);
-	cancel_delayed_work_sync(&hdp->hdcp.check_work);
+
+	if (hdp->hdcp.enable == 1)
+		cancel_delayed_work_sync(&hdp->hdcp.check_work);
+
+	hdp->hdcp.enable = 0;
 
 	return ret;
 }
-- 
2.17.1

